package/MDAnalysis/lib/_cutil.pyx

```diff
@@ -47,47 +59,951 @@ ctypedef cset[int] intset
 ctypedef cmap[int, intset] intmap
 
 
-@cython.boundscheck(False) # turn off bounds-checking for entire function
-@cython.wraparound(False)  # turn off negative index wrapping for entire function
-def unique_int_1d(np.int64_t[:] values):
+cdef inline bint isinf(double x) nogil:
+    """Check if a double is ``inf`` or ``-inf``.
+
+    In contrast to ``libc.math.isfinite``, this function keeps working when
+    compiled with ``-ffast-math``.
+
+
+    .. versionadded:: 0.20.0
+    """
+    return (x == INFINITY) | (x == -INFINITY)
+
+
+def coords_add_vector(float[:, :] coordinates not None,
```

Comment:
I don't really get this, but I guess I need to find the time to play with it myself.  I don't like the idea that we're implementing what should already be done optimally in numpy.  If we start having functions like this, then the library could quickly balloon with similar....

Reply:
I absolutely agree that this *should* be done with numpy. However, the current implementation in numpy is dead slow (about a factor 10 slower than it could be) due of the way numpy implements broadcasting in its vectorized additions in `sse2_binary_add_FLOAT()`, resulting in a gazillion of `__memmove_ssse3()`s. I should probably think about a more efficient way of implementing this in numpy instead of implementing a workaraound in MDAnalysis...

Reply:
+1 for moving upstream instead of here if there's some fundamental flaw in NumPy operations that can be changed without drawbacks

Reply:
@tylerjereddy I'm currently investigating where and how shortcuts in/around numpy's `npyiter_copy_to_buffers()` might be implemented. Unfortunately, understanding this section of the numpy code base is not an entirely trivial task...

Reply:
@zemanj I went ahead and tried to summarize your idea in the [weekly NumPy meeting community topics section](https://hackmd.io/WVHUbdriRe26t6s09HjEhA?view). I hope that's ok & you are welcome to join / listen in on the call using the link on that doc.

Feel free to clean up my summary there too--anyone can edit that.

The call is in a few hours--12 pm Pacific, which may be rather late for you.

Reply:
@tylerjereddy Thanks! I added two links, one to the gdb log and one to the call stack visualization I posted in the discussion of #2210. I won't be able to join the meeting today, so feel free to postpone the topic.

Reply:
This part of numpy is really complex and probably can be improved. And yes it is complex...

The buffer code should skip copying to the buffer when the correct data is already in it (at least in most cases). But I am not sure it is very smart about trying to ensure that this actually happens. I have to check the exact case you are observing, possibly there are also some cases which could dispatch to a non-buffered path, but do not?